proposals/shared-everything-threads/Overview.md

```diff
@@ -194,37 +107,260 @@ how a non-shared function accesses a `shared` memory):
+globals for a module in the map and use pairs of instance and thread IDs as the keys instead of
+triples. Thread-local global accesses would then lower to accesses to the hash map to get the TLS
+block base address followed by applications of an offset to access the global within the block. This
+implementation has the advantage that the TLS block base address can be reused across accesses to
+multiple thread-local globals, and in the extreme could be pinned to a register to eliminate all but
+the initial hash map access for each instance on each thread.
+
+This design also allows implementations to be flexible regarding when they initialize thread-local
+globals. An engine wishing to eagerly initialize the globals would initialize the new TLS for every
+instance in the system every time a new thread is created and would initialize the new TLS for every
+thread in the system every time a new instance is created. Alternatively, an implementation could
+lazily initialize the TLS for an instance on a thread the first time a function from that instance
+(or any other instance that has imported a TLS global from that instance) is executed, ensuring that
+globals are initialized before they can be accessed without spending time initializing globals that
+will never be accessed.
+
+### Thread Management Builtins
+
+Standalone WebAssembly engines (i.e., non-browser) do not have web workers as a parallel primitive.
+For these environments, we propose to modify the [component model] by adding the following builtins
+to allow creating and managing threads:
+
+```
+canon ::= ...
+          | (canon thread.spawn <typeidx> (core func <id>?))
+          | (canon thread.hw_concurrency (core func <id>?))
+```
+
+These new builtins would join the short list of current [canonical builtins]: `lift`, `lower`, and
+`resource.*` management. The new `thread.*` builtins have as a goal to be polyfillable on the web.
+
+[canonical builtins]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#canonical-abi
+
+#### `thread.spawn`
+
+The `thread.spawn` built-in has type `[f: (ref null $f) n:i32 c:i32] -> []`, where `$f` is `(rec
+(type (sub final (func shared (param i32))))).0`: it invokes the function `f` `n` times, potentially
+in parallel, passing `c` to each invocation.
+
+The need to pass a function reference is discussed [here][static-function-discussion]. Note that,
+though the component model ABI does not allow passing `funcref`s, the canonical builtins have no
+such restriction.
+
+Ideally we would allow thread entry points to take arbitrary parameters, as discussed
+[here][function-type-discussion]. Until we get type parameters in core Wasm, however, that would
+require a name mangling scheme to describe the intended type of the import. It's unclear that there
+can be any satisfactory name mangling scheme that would handle arbitrary GC types. For now we avoid
+that complexity by using a single concrete type.
```

Comment:
Until we have type parameters, supporting more than one static type for the callback would require multiple imports. In principle we could use the same import name for multiple imports of different types, but that would mean the import provider cannot be implemented in Wasm. We would therefore want to use different names, i.e. a mangling scheme, to differentiate between different types. Even then, a particular provider implemented in Wasm could only provide a finite set of imports, though.

I'm not sure whether the component model folks are ok with any amount of name mangling, which is why we opted to support just a single signature for now.

Reply:
I agree that we should avoid import overloading. But I'm not sure how the problem here is different from the situation for lift and lower. Is it because they are not expected to be compiled to an auxiliary import?

Reply:
Hmm, good question. I'd have to defer to someone who knows more about the component model than I do for that one. @lukewagner, do you have thoughts about this?

Reply:
There's not a fundamental problem (for the reasons Andreas is thinking); it's more of a complexity that emerges as a practical toolchain implementation detail.  Basically, we want the core wasm producer toolchain (e.g., clang+ld) to not have to deeply know about components; they want to simply spit out a core module that is then wrapped into a component by a tool like [`wit-component`](https://crates.io/crates/wit-component) as the final step of the build process.  What `wit-component` wants to do is inspect all the imports of the incoming core module and emit `canon` definition that satisfies them based just on the name.  Of course it's possible to smuggle in extra metadata to `wit-component` via a custom section and indeed this is done for the target `world` (which otherwise `wit-component` could not infer from core import/export names/types).  But `thread.spawn` doesn't appear in the `world` (that's the point).

FWIW, a fairly easy solution (that doesn't involve mangling an arbitrary core wasm type into the function name, which is what I think the initial concern was about) is to say that `wit-component` will infer the closure type of `thread.spawn` from the imported core function signature (requiring the two places that the closure type appears in the overall function signature to be identical to avoid unification complexities).  Then the only question is what is the core import name that `wit-component` will look for.  Given that it's not going away, I think it'd be "fine" to abuse import overloading and just fix a single name for all `thread.spawn` core imports, but if we wanted to avoid import overloading, we could almost as easily allow an ignored suffix that the core toolchain could then use to avoid duplicates by appending `-1`, `-2` or whatever.  But the import overloading might be meaningfully easier (when you consider separate compilation) and thus worth it, if it doesn't blow up some part of the toolchain.

So overall I'm ambivalent: I think it wouldn't be terrible to remove this restriction, but I can appreciate not wanting to even have to think about this in a first release.

Reply:
So is virtualizability of the thread spawn imports not a concern because they are builtins, and therefore not subject to the normal design goals of normal component interfaces?

Reply:
That's right, yes.  Component built-ins (`thread.spawn`, `resource.new`, `resource.drop`, ...) are like instructions in wasm; you don't need to import them from the outside.